### PR TITLE
#1Task. ivan.lozovskij

### PR DESCRIPTION
**Pull request by Lozovskij**
https://travis-ci.org/github/Lozovskij/db-assignments/builds/711217799

### CODE REVIEW

File: task/sql-tasks.js [hunk 1]
```diff
@@ -43,7 +43,16 @@ async function task_1_1(db) {
  *
  */
 async function task_1_2(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    SELECT 
+        OrderID as "Order Id", 
+        SUM(Quantity * UnitPrice) as "Order Total Price",
+        ROUND( SUM(Discount * Quantity) / SUM(UnitPrice * Quantity) * 100, 3) as "Total Order Discount, %"
+    FROM northwind.OrderDetails
```
Comment (AlexMokin) [inline task/sql-tasks.js]: Do not use schema name in the queries. It should be part of the connection settings. Fix in the whole file

File: task/sql-tasks.js [hunk 2]
```diff
@@ -106,7 +143,18 @@ async function task_1_6(db) {
  *
  */
 async function task_1_7(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    SELECT 
+        table_1.EmployeeID as EmployeeId, 
+        CONCAT(table_1.FirstName, ' ', table_1.LastName) as FullName,
+        CASE
```
Comment (AlexMokin) [inline task/sql-tasks.js]: see ifnull and coalesce functions

File: task/sql-tasks.js [hunk 3]
```diff
@@ -178,7 +260,13 @@ async function task_1_12(db) {
  *
  */
 async function task_1_13(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    SELECT
+        COUNT(Discontinued) as TotalOfCurrentProducts,
+        SUM(CASE Discontinued WHEN 1 THEN 1 ELSE 0 END) as TotalOfDiscontinuedProducts
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use count when you need to count recrods to show intent explicitly

Comment (AlexMokin) [reply task/sql-tasks.js]: but even bettter to write this query like this

```
select
(select count()...) as 'T'
(select count()...) as 'TT'
```

Notice that from-clause is optional

File: task/sql-tasks.js [hunk 4]
```diff
@@ -200,7 +292,23 @@ async function task_1_14(db) {
  *
  */
 async function task_1_15(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    SELECT 
+    SUM(CASE YEAR(OrderDate) WHEN 1997 THEN (CASE month(OrderDate) WHEN 1 THEN 1 ELSE 0 END) ELSE 0 END) as January,
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use count when you need to count recrods to show intent explicitly

File: task/sql-tasks.js [hunk 5]
```diff
@@ -200,7 +292,23 @@ async function task_1_14(db) {
  *
  */
 async function task_1_15(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    SELECT 
+    SUM(CASE YEAR(OrderDate) WHEN 1997 THEN (CASE month(OrderDate) WHEN 1 THEN 1 ELSE 0 END) ELSE 0 END) as January,
+    SUM(CASE YEAR(OrderDate) WHEN 1997 THEN (CASE month(OrderDate) WHEN 2 THEN 1 ELSE 0 END) ELSE 0 END) as February,
```
Comment (AlexMokin) [inline task/sql-tasks.js]: check year in where clause

File: task/sql-tasks.js [hunk 6]
```diff
@@ -236,7 +358,16 @@ async function task_1_17(db) {
...
 async function task_1_18(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    SELECT * 
+    FROM (SELECT 
+            date_format(OrderDate, '%Y-%m-%d %T') as OrderDate, 
+            COUNT(OrderID) as "Total Number of Orders" 
+            FROM northwind.Orders 
+            GROUP BY OrderDate) AS T
+    WHERE YEAR(T.OrderDate) = 1998;
```
Comment (AlexMokin) [inline task/sql-tasks.js]: implement this without using subqueries

File: task/sql-tasks.js [hunk 7]
```diff
@@ -281,7 +446,37 @@ async function task_1_21(db) {
  * @return {array}
  */
 async function task_1_22(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    SELECT t1.CompanyName, t1.ProductName, t1.PricePerItem
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use only 1 subquery in this task

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

File: task/sql-tasks.js [hunk 8]
```diff
@@ -200,7 +288,22 @@ async function task_1_14(db) {
  *
  */
 async function task_1_15(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    SELECT
```
Comment (AlexMokin) [inline task/sql-tasks.js]: implement without using subqueryies. Check year only once in where-clause

File: task/sql-tasks.js [hunk 9]
```diff
@@ -281,7 +439,27 @@ async function task_1_21(db) {
...
+    let result = await db.query(`
+    SELECT DISTINCT
+		Customers.CompanyName as CompanyName,
+		Products.ProductName as ProductName,
+		OrderDetails.UnitPrice as PricePerItem
+	FROM Orders
+		INNER JOIN OrderDetails on Orders.OrderID = OrderDetails.OrderID
+		INNER JOIN Customers on Orders.CustomerID = Customers.CustomerID
+		INNER JOIN Products on OrderDetails.ProductID = Products.ProductID
+	WHERE OrderDetails.UnitPrice =
```
Comment (AlexMokin) [inline task/sql-tasks.js]: this is kinda slow. In subquery, try to group by customer id and join subquery on price and customer id

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

File: task/sql-tasks.js [hunk 10]
```diff
@@ -281,7 +441,29 @@ async function task_1_21(db) {
...
  */
 async function task_1_22(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    SELECT DISTINCT
+		Customers.CompanyName as CompanyName,
+		Products.ProductName as ProductName,
+		OrderDetails.UnitPrice as PricePerItem
+	FROM Customers 
+		RIGHT JOIN Orders on Orders.CustomerID = Customers.CustomerID
```
Comment (vramaniuk) [inline task/sql-tasks.js:450]: `RIGHT JOIN` is redundant because we have no orders without customers, just JOIN

File: task/sql-tasks.js [hunk 11]
```diff
@@ -281,7 +441,29 @@ async function task_1_21(db) {
...
+        RIGHT JOIN OrderDetails on Orders.OrderID = OrderDetails.OrderID
```
Comment (vramaniuk) [inline task/sql-tasks.js:451]: ` RIGHT JOIN` is redundant because we have no orderdetails without orders , just JOIN

**Review by vramaniuk — COMMENTED** [review]
Closing the PR as accepted! Thanks!

**Pull request closed without merging**